### ISSUE TITLE
PlaceholderFactory's Create method parent transform overload

### ISSUE THREAD

**Issue by HDANILO** [ssannandeji/Zenject-2019#557]
Is it possible to add an overload to the PlaceholderFactory's create method to accept a transform, and when instantiating the GameObject, instantiate directly to the transform as it's parent?

I know there is the UnderTransform binding rule, but sometimes we need to instantiate prefabs on different parents, even more if we don't have Nested Prefabs(might be coming soon)

**Comment by cooloon** [comment]
Passing transform as a parameter doesn't suit for your case?
```
public class SomeFactory : Zenject.PlaceholderFactory<Transform, SomeClass> { }
```

Or can't you parent transform after instance is created?
```
var instance = factory.Create();
instance.SetParent(parent);
```

(If instance is created in sub context, you have to manage instance lifecycle by yourself.)

**Comment by svermeulen** [comment]
Yeah I would also suggest just having transform be another parameter.  Then you could use a custom factory like this:

```csharp
public class Foo : MonoBehaviour
{
    public class Factory : PlaceholderFactory<Transform, Foo>
    {
    }
}

public class FooFactory : IFactory<Transform, Foo>
{
    readonly IFactory<Foo> _actualFactory;

    public FooFactory(IFactory<Foo> actualFactory)
    {
        _actualFactory = actualFactory;
    }

    public Foo Create(Transform parent)
    {
        var foo = _actualFactory.Create();

        if (parent != null)
        {
            foo.transform.SetParent(parent, false);
        }

        return foo;
    }
}

public class TestInstaller : MonoInstaller
{
    public override void InstallBindings()
    {
        Container.BindFactory<Transform, Foo, Foo.Factory>().FromFactory<FooFactory>();
        Container.BindIFactory<Foo>().FromNewComponentOnNewGameObject().WhenInjectedInto<FooFactory>();
    }
}

```

You could inject DiContainer into your custom factory and then call DiContainer.InstantiateX methods directly, but using another factory like this is better because then Foo is valiidated.

**Comment by HDANILO** [comment]
Both answer works, but it's not as good as instantiating directly on the parent transform.

The first advantage on instantiating directly on the parent is the performance, moving/reparenting objects on the hierarchy is expensive.

Second problem is that it can cause some misbehaviour on the transform, the reason is because when you create the object without a parent, it'll go to your root hierarchy, which is probably a Transform, if you reparent it on a UI object, thus, a RectTransform, the scale, position and rotation might be weird.

I think those are arguments enough to have PlaceholdFactory's create method an overload to a parent so that everybody can easily specify a transform to be used as parent in the instantiate process.

**Comment by svermeulen** [comment]
Yeah I can see why you might want to instantiate directly on the parent.  But you don't need an overload on PlaceholderFactory to do this.  You can use my above solution except directly call DiContainer.InstantiateX and forward the parent to that.

**Comment by svermeulen** [comment]
Though, if you do go that route you might want to also implement IValidatable manually as explained in the docs

**Comment by HDANILO** [comment]
Sure, but I think it is such a common use case that could be directly handled by the Zenject. Ofc Zenject already provide enough support to implement any solution, it's just that It make so much more sense to have it directly on the PlaceholdFactory than adding an extra argument to the PlaceholdFactory and having to make concrete implementation of a factory

**Comment by HDANILO** [comment]
The whole idea of PlaceholdFactory is very genius. Having to make an IFactory everytime I want to make a dynamic parent just kills a little bit of that geniousity.

**Comment by svermeulen** [comment]
PlaceholderFactory is often not even used for MonoBehaviours so an overload that takes an optional transform parent doesn't make sense.  To me, an optional transform parent makes a lot more sense as just another parameter.

If you're concerned that you're going to need this functionality in a lot of places, I'd suggest making a generic factory from FooFactory above that works with any MonoBehaviour and then you can avoid the boilerplate code every time.

**Comment by HDANILO** [comment]
That makes sense to me, what if we make a GameObjectFactory as an alternative to the PlaceholderFactory, but ofc, only for GameObjects? This way we could have this functionality already integrated in.

Ofc I can make it internally on our project, but It would be interesting to know if anyone else on the community would be interested in having this already in the Zenject, this way we can have it standardized.

**Comment by HDANILO** [comment]
Also, handling such implementation to be really generic can be very cumbersome. Imagine that besides that parent, a gameobject might need more parameters, such things are already handled very well by PlaceholdFactories, and an IFactory with Container.Instantiate might not just do the job, we might need to use sub container for such case, but right now I'm not even sure if only subcontainers would do the job for all cases. Also, not sure how much optimal that solution would be in terms of performance(talking about all reflections and everything).

**Comment by svermeulen** [comment]
Yeah you're right that making a truly generic version of that factory wouldn't be great, since you'd need to have a different one for each DiContainer.InstantiateX method.  You can pass parameter lists to DiContainer.InstantiateX methods though so it would be doable, but yes I admit it's not a great solution.

By the way, I thought of a clever way to add your overload method to only PlaceholderFactory's that use Component types by using extension methods:

```csharp
public static class PlaceholderFactoryExtensions
{
    public static TValue Create<TValue>(this PlaceholderFactory<TValue> factory, Transform parent)
        where TValue : Component
    {
        var value = factory.Create();

        if (parent != null)
        {
            value.transform.SetParent(parent, false);
        }

        return value;
    }

    public static TValue Create<TParam1, TValue>(this PlaceholderFactory<TParam1, TValue> factory, TParam1 p1, Transform parent)
        where TValue : Component
    {
        var value = factory.Create(p1);

        if (parent != null)
        {
            value.transform.SetParent(parent, false);
        }

        return value;
    }

    // .. etc.
}
```

Of course, it doesn't support directly using the parent to save some performance, but it's a pretty nice and easy solution at least.

**Comment by cooloon** [comment]
Yeah uGUI is ugly. Specially when instantiating prefab under RectTransform.

Some workarounds are discussed in Unity forum. Setting sizeDelta, anchoredPosition, or localScale after parenting can help.
https://forum.unity.com/threads/problem-with-instantiate-ui-prefabs.271345/

But it seems to be out of Zenject scope.

**Comment by JonMcPherson** [comment]
+1 to being able to create factory bindings that instantiate directly on the parent transform rather than having to `SetParent` after the object is already instantiated and in the scene.

Maybe an alternative to `UnderTransform` that can configure the binding to use
```
UnityEngine.Object.Instantiate(prefabObject, parentTransform, false)
```
(and the other overloads).
